### PR TITLE
feat(amf): Encoding and Decoding of session modification IEs

### LINKED ISSUE #12439: [AMF]: Encoding & Decoding of session modification IEs
As part of session modification feature for 5G few IEs needs to be added. 
These are part of PR : https://github.com/magma/magma/pull/11989/

As this PR is little bit bigger, its better to move the IE encoding/decoding part out of this PR and track it separately. 

Changes will be  :
1. Encoding of the session modification library as part of nas5g
2. Decoding of session modification library as part of nas5g
3.  Unit testing around that.

### PR DESCRIPTION
**Pull request by sreedharkumartn**
Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

## Summary
- As part of session modification feature for 5G few IEs needs to be added. These are part of PR : #11989. As this PR is little bit bigger, raised this PR for session modification IEs.

- Changes made:
    - Encoding of the session modification library as part of nas5g
    - Decoding of session modification library as part of nas5g
    - Unit testing around that.

- Newly Added NAS5G files:
    - ``lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationCommand.cpp``
    - ``lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationCommandReject.cpp``
    - ``lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationComplete.cpp``
    - ``lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GQosFlowDescriptor.cpp``
    - ``lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GQosFlowParam.cpp``

## Test Plan
- Validated with UT.
UT

- Done basic sanity with UERANSIM.
PCAP Snap:
image
image

mme log:
[mme.log](https://github.com/magma/magma/files/8489801/mme.log)

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Review by nstng — APPROVED** [review]
Bazel changes LGTM - build and test run through on CI.

File: lte/gateway/c/core/oai/tasks/nas5g/include/M5GPDUSessionModificationCommand.hpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/*
+   Copyright 2020 The Magma Authors.
```
Comment (LKreutzer) [inline lte/gateway/c/core/oai/tasks/nas5g/include/M5GPDUSessionModificationCommand.hpp]: For new files the year in the licence header should be the current year "2022".

File: lte/gateway/c/core/oai/tasks/nas5g/include/M5GPDUSessionModificationCommandReject.hpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/*
+   Copyright 2020 The Magma Authors.
```
Comment (LKreutzer) [inline lte/gateway/c/core/oai/tasks/nas5g/include/M5GPDUSessionModificationCommandReject.hpp]: Same as above.

File: lte/gateway/c/core/oai/tasks/nas5g/include/M5GPDUSessionModificationComplete.hpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
+/*
+   Copyright 2020 The Magma Authors.
```
Comment (LKreutzer) [inline lte/gateway/c/core/oai/tasks/nas5g/include/M5GPDUSessionModificationComplete.hpp]: Same as above.

File: lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GGprsTimer.hpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2020 The Magma Authors.
```
Comment (LKreutzer) [inline lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GGprsTimer.hpp]: Same as above.

File: lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GQosFlowDescriptor.hpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
+/*
+   Copyright 2020 The Magma Authors.
```
Comment (LKreutzer) [inline lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GQosFlowDescriptor.hpp]: Same as above.

File: lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GQosFlowParam.hpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
+/*
+   Copyright 2020 The Magma Authors.
```
Comment (LKreutzer) [inline lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GQosFlowParam.hpp]: Same as above.

File: lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationCommand.cpp [hunk 7]
```diff
@@ -0,0 +1,226 @@
+/*
+   Copyright 2020 The Magma Authors.
```
Comment (LKreutzer) [inline lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationCommand.cpp]: Same as above.

File: lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationCommandReject.cpp [hunk 8]
```diff
@@ -0,0 +1,146 @@
+/*
+   Copyright 2020 The Magma Authors.
```
Comment (LKreutzer) [inline lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationCommandReject.cpp]: Same as above.

File: lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationComplete.cpp [hunk 9]
```diff
@@ -0,0 +1,113 @@
+/*
+   Copyright 2020 The Magma Authors.
```
Comment (LKreutzer) [inline lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationComplete.cpp]: Same as above.

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GGprsTimer.cpp [hunk 10]
```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 020 The Magma Authors.
```
Comment (LKreutzer) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GGprsTimer.cpp]: Typo here - but still same as above.

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GQosFlowDescriptor.cpp [hunk 11]
```diff
@@ -0,0 +1,83 @@
+/*
+   Copyright 2020 The Magma Authors.
```
Comment (LKreutzer) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GQosFlowDescriptor.cpp]: Same as above.

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GQosFlowParam.cpp [hunk 12]
```diff
@@ -0,0 +1,252 @@
+/*
+   Copyright 2020 The Magma Authors.
```
Comment (LKreutzer) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GQosFlowParam.cpp]: Same as above.

**Review by LKreutzer — CHANGES_REQUESTED** [review]
See comments regarding the year in the licence headers. The bazel changes lgtm.

File: lte/gateway/c/core/oai/tasks/nas5g/include/M5GPDUSessionModificationCommand.hpp (hunk 1 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/include/M5GPDUSessionModificationCommand.hpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/include/M5GPDUSessionModificationCommandReject.hpp (hunk 2 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/include/M5GPDUSessionModificationCommandReject.hpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/include/M5GPDUSessionModificationComplete.hpp (hunk 3 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/include/M5GPDUSessionModificationComplete.hpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GGprsTimer.hpp (hunk 4 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GGprsTimer.hpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GQosFlowDescriptor.hpp (hunk 5 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GQosFlowDescriptor.hpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GQosFlowParam.hpp (hunk 6 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GQosFlowParam.hpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationCommand.cpp (hunk 7 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationCommand.cpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationCommandReject.cpp (hunk 8 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationCommandReject.cpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationComplete.cpp (hunk 9 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/M5GPDUSessionModificationComplete.cpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GGprsTimer.cpp (hunk 10 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GGprsTimer.cpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GQosFlowDescriptor.cpp (hunk 11 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GQosFlowDescriptor.cpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GQosFlowParam.cpp (hunk 12 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GQosFlowParam.cpp]: Addressed Review Comment

**Review by LKreutzer — APPROVED** [review]
Bazel changes and licences lgtm!

**Pull request closed without merging**